### DISCUSSION TITLE
Unable to load shared library 'Magick.Native-Q8-x64.dll' or one of its dependencies.

### DISCUSSION THREAD

**Discussion by ajaykodam** [dlemstra/Magick.NET#1314] [Help] [answered]
I am using AWS Lambda with .net core V6.0, I have one nuget package `test.common` where that nuget package using the `Magick.NET-Q8-AnyCPU V12.2.1`. I have referenced `test.common` nuget package into `test.lambda` project. Using docker file I am generating the docker image and deploying that image to AWS Lambda.

Here is my code reside in the `test.common` nuget package where I am generating the jpg image from given text. And this method I am using it in `test.lambda` project

```
private byte[] GetImageDataFromText(string text)
        {
            byte[] data = Encoding.UTF8.GetBytes(text);
            using (MagickImage image = new MagickImage(data, MagickFormat.Text))
            {
                image.Format = MagickFormat.Jpeg;

                MemoryStream outputMemoryStream = new();

                image.Write(outputMemoryStream);

                return outputMemoryStream.ToArray();
            }
        }
```

Here is the docker file

```
# FROM public.ecr.aws/lambda/dotnet:6 AS base
FROM amazon/aws-lambda-dotnet:6 AS base

FROM mcr.microsoft.com/dotnet/sdk:latest AS build

WORKDIR /src
ARG awstoken
ARG awsaccountid
ARG awsregion

COPY ./test.Lambda/ ./test.Lambda/

RUN dotnet nuget add source "https://text-$awsaccountid.d.codeartifact.$awsregion.amazonaws.com/nuget/testCompany/v3/index.json" -u aws -p "$awstoken" --store-password-in-clear-text

FROM build AS publish
RUN dotnet publish "test.Lambda/test.Lambda.csproj" -c Release -o /app/publish

RUN apt-get update \
    && apt-get install -y --no-install-recommends \
        fontconfig \
        libopenjp2-7 \
        libfontconfig1 \
    && rm -rf /var/lib/apt/lists/

FROM base AS final
WORKDIR /var/task  
COPY --from=publish /app/publish . 
CMD ["test.Lambda::test.Lambda.Function::FunctionHandler"]
```
I have added properties to my .csprj file as well

  ```
<CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
    <MagickCopyNativeWindows>true</MagickCopyNativeWindows>
    <MagickCopyNativeLinux>true</MagickCopyNativeLinux>
    <MagickCopyNativeLinuxMusl>true</MagickCopyNativeLinuxMusl>
    <MagickCopyNativeMacOS>true</MagickCopyNativeMacOS>
```

When I execute the my `test.lambda` I am getting the error

`The type initializer for 'NativeMagickSettings' threw an exception.\nInner Exception: Unable to load shared library 'Magick.Native-Q8-x64.dll' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libMagick.Native-Q8-x64.dll: cannot open shared object file: No such file or directory`

**Let me know if anything missing here, as this is critical for my project deadline, Please help ASAP**

**Comment by dlemstra** [comment]
It looks like that docker image is using an old operating system that is not yet using glibc 2.27:

```
bash-4.2# ldd native/Magick.Native-Q8-x64.dll.so
ldd: warning: you do not have execution permission for `native/Magick.Native-Q8-x64.dll.so'
native/Magick.Native-Q8-x64.dll.so: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by native/Magick.Native-Q8-x64.dll.so)
native/Magick.Native-Q8-x64.dll.so: /lib64/libc.so.6: version `GLIBC_2.27' not found (required by native/Magick.Native-Q8-x64.dll.so)
        linux-vdso.so.1 (0x00007ffcb6df3000)
        libresolv.so.2 => /lib64/libresolv.so.2 (0x00007f48ee166000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f48edf48000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f48edc08000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f48ed85b000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f48f09ce000)
```

**Reply by ajaykodam** [reply]
@dlemstra Can you please let me know if I use latest os version which one should I use?

**Comment by dlemstra** [comment]
I ran some experiments and it looks like I managed to link the code with an older version of `glibc` to make it work on your container image. I now get the following output:

```
bash-4.2# ldd Magick.Native-Q8-x64.dll.so
        linux-vdso.so.1 (0x00007ffcfd5a7000)
        libresolv.so.2 => /lib64/libresolv.so.2 (0x00007fac7957c000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fac7935e000)
        libm.so.6 => /lib64/libm.so.6 (0x00007fac7901e000)
        libc.so.6 => /lib64/libc.so.6 (0x00007fac78c71000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fac7bde9000)
```

But I haven't ran any unit test to make sure this works. I will keep you updated.

**Reply by ajaykodam** [reply]
@dlemstra Thank you for fixing in older version, Can I use it now or do I need to wait for new version?

**Reply by dlemstra** [reply]
You will need to wait for a new version to published.

**Reply by ajaykodam** [reply]
@dlemstra Thank you!!! When can I expect the new version?

**Reply by dlemstra** [reply]
I don't plan my releases but I will try to publish a new release next week.

**Reply by ajaykodam** [reply]
@dlemstra  Ok Thanks

**Reply by ajaykodam** [reply]
@dlemstra Please let me know when can I get the new version of the fix, as we dragging the production deployment for this.

**Reply by dlemstra** [reply]
The new release has been published.

**Reply by ajaykodam** [reply]
@dlemstra Thank you so much!!! I will check with latest nuget package update and let you know.

**Comment by ajaykodam** [comment, marked as answer]
@dlemstra After I updated the latest nuget package I am getting the below error, Can you please help me?

`Fontconfig error: Cannot load default config file
`
`ImageMagick.MagickTypeErrorException: UnableToReadFont `helvetica' @ error/annotate.c/RenderFreetype/1619\n   at ImageMagick.NativeInstance.CheckException(IntPtr exception, IntPtr result) in /_/src/Magick.NET/Native/NativeInstance.cs:line 57\n   at ImageMagick.MagickImage.NativeMagickImage.ReadBlob(IMagickSettings`1 settings, Byte[] data, Int32 offset, Int32 length) in /_/src/Magick.NET/Native/MagickImage.cs:line 7069\n   at ImageMagick.MagickImage.Read(Byte[] data, Int32 offset, Int32 length, IMagickReadSettings`1 readSettings, Boolean ping, String fileName) in /_/src/Magick.NET/MagickImage.cs:line 7639\n   at ImageMagick.MagickImage.Read(Byte[] data, MagickFormat format) in /_/src/Magick.NET/MagickImage.cs:line 4791\n   at ImageMagick.MagickImage..ctor(Byte[] data, MagickFormat format) in /_/src/Magick.NET/MagickImage.cs:line 97\n `

**Reply by dlemstra** [reply]
Looks like you need to install fontconfig in your container.

**Reply by ajaykodam** [reply]
I am already running this in docker file, see above on this discussion
```

RUN apt-get update \
    && apt-get install -y --no-install-recommends \
        fontconfig \
        libopenjp2-7 \
        libfontconfig1 \
    && rm -rf /var/lib/apt/lists/
```

**Reply by dlemstra** [reply]
You probably also need to install fonts?

**Reply by ajaykodam** [reply]
Currently I am doing this but still no luck.

```
RUN apt-get update \
    && apt-get upgrade -y \
    && apt-get install -y --no-install-recommends \
        fontconfig \
        libopenjp2-7 \
        libfontconfig1 \
    && rm -rf /var/lib/apt/lists/

RUN sed -i'.bak' 's/$/ contrib/' /etc/apt/sources.list
RUN apt-get update \
    && apt-get install -y ttf-mscorefonts-installer fontconfig

RUN fc-cache -f -v
```

Can you please let me know which fonts I need to install for linux?

`Fontconfig error: Cannot load default config file`

Do I need to set explicitly specify any environment path in my docker file?

**Reply by ajaykodam** [reply]
Please help me on this? 
Where can I get this for linux?

`UnableToReadFont `helvetica' @ error/annotate.c/RenderFreetype/1619\n`

![image](https://user-images.githubusercontent.com/120182902/208098979-d487965c-ae3e-4ea3-a66e-7b23e1e8ad92.png)

Is this code causing?

**Reply by dlemstra** [reply]
This looks like it is related to your docker configuration and not the code.

**Reply by ajaykodam** [reply]
Could you please help me what is missing here?

```
# FROM public.ecr.aws/lambda/dotnet:6 AS base
FROM amazon/aws-lambda-dotnet:6 AS base

FROM mcr.microsoft.com/dotnet/sdk:latest AS build

WORKDIR /src
ARG awstoken
ARG awsaccountid
ARG awsregion

COPY ./test.Lambda/ ./test.Lambda/

RUN dotnet nuget add source "https://text-$awsaccountid.d.codeartifact.$awsregion.amazonaws.com/nuget/testCompany/v3/index.json" -u aws -p "$awstoken" --store-password-in-clear-text

FROM build AS publish
RUN dotnet publish "test.Lambda/test.Lambda.csproj" -c Release -o /app/publish
RUN apt-get update \
    && apt-get upgrade -y \
    && apt-get install -y --no-install-recommends \
        fontconfig \
        libopenjp2-7 \
        libfontconfig1 \
        libwmf0.2-7  \
    && rm -rf /var/lib/apt/lists/

RUN fc-cache -f -v

FROM base AS final
ENV FONTCONFIG_PATH=/etc/fonts
WORKDIR /var/task  
COPY --from=publish /app/publish . 
CMD ["test.Lambda::test.Lambda.Function::FunctionHandler"]
```

**Reply by ajaykodam** [reply]
@dlemstra From this error message I think you know what missing in my docker file?

`ImageMagick.MagickTypeErrorException: UnableToReadFont `helvetica' @ error/annotate.c/RenderFreetype/1619\n   at ImageMagick.NativeInstance.CheckException(IntPtr exception, IntPtr result) in /_/src/Magick.NET/Native/NativeInstance.cs:line 57\n   at ImageMagick.MagickImage.NativeMagickImage.ReadFile(IMagickSettings`1 settings) in /_/src/Magick.NET/Native/MagickImage.cs:line 7133\n   at ImageMagick.MagickImage.Read(String fileName, IMagickReadSettings`1 readSettings, Boolean ping) in /_/src/Magick.NET/MagickImage.cs:line 7690\n   at ImageMagick.MagickImage.Read(String fileName, IMagickReadSettings`1 readSettings) in /_/src/Magick.NET/MagickImage.cs:line 4941\n   at ImageMagick.MagickImage..ctor(String fileName, IMagickReadSettings`1 readSettings) in /_/src/Magick.NET/MagickImage.cs:line 220\n   a`

**Reply by dlemstra** [reply]
To get rid of the warning `Fontconfig error: Cannot load default config file` you can do this:

```C#
MagickNET.SetFontConfigDirectory("/etc/fonts");
```

I found that folder with the following command:

```
bash-4.2# find / -name fonts.conf
/etc/fonts/fonts.conf
```

And you are getting that exception because your docker image has no `helvetica` font installed. You can see the fonts that are installed like this:
```
bash-4.2# fc-list
/usr/share/fonts/dejavu/DejaVuSansCondensed-Oblique.ttf: DejaVu Sans,DejaVu Sans Condensed:style=Condensed Oblique,Oblique
/usr/share/fonts/dejavu/DejaVuSansCondensed-Bold.ttf: DejaVu Sans,DejaVu Sans Condensed:style=Condensed Bold,Bold
/usr/share/fonts/dejavu/DejaVuSans.ttf: DejaVu Sans:style=Book
/usr/share/fonts/dejavu/DejaVuSans-Bold.ttf: DejaVu Sans:style=Bold
/usr/share/fonts/dejavu/DejaVuSansCondensed.ttf: DejaVu Sans,DejaVu Sans Condensed:style=Condensed,Book
/usr/share/fonts/dejavu/DejaVuSans-ExtraLight.ttf: DejaVu Sans,DejaVu Sans Light:style=ExtraLight
/usr/share/fonts/dejavu/DejaVuSansCondensed-BoldOblique.ttf: DejaVu Sans,DejaVu Sans Condensed:style=Condensed Bold Oblique,Bold Oblique
/usr/share/fonts/dejavu/DejaVuSans-Oblique.ttf: DejaVu Sans:style=Oblique
/usr/share/fonts/dejavu/DejaVuSans-BoldOblique.ttf: DejaVu Sans:style=Bold Oblique
```

As you can see only the `DejaVu Sans` font seems to be installed on the system. You could decide to use that font like this:

```C#
// With a font name
var settings = new MagickReadSettings
{
    FontFamily = "DejaVu Sans"
};

// Or a path to a font file (I would advise this option)
var settings =  new MagickReadSettings
{
    Font = "/var/task/YourFont.ttf"
};

using var image = new MagickImage("label:I think you are doing this somewhere in your code", settings);
```

**Reply by ajaykodam** [reply]
@dlemstra Thank you so much!! I will experiment as you mentioned.

**Reply by ajaykodam** [reply]
@dlemstra This works for me, as I have to give exact `.ttf` file path
```
// Or a path to a font file (I would advise this option)
var settings =  new MagickReadSettings
{
    Font = "exact file path"
};
```

**Comment by ajaykodam** [comment]
@dlemstra Currently we are checking the platform-specific condition to locate the correct font path if it's Linux/Unix platform. If it's windows as we don't need to set this it will take the default font of the system. And we don't know when the version of Linux/Unix stops supporting the Deja Vu font. Do you have a way to incorporate these changes in your NuGet packages?

Currently I am doing this.

```
if (Environment.OSVersion.Platform == PlatformID.Unix)
  {
      readSettings.Font = "/usr/share/fonts/dejavu/DejaVuSans.ttf";
  }
```

**Reply by dlemstra** [reply]
You could also include the font file in your .NET project and copy it to the output and then use the file path in your container and then you don't need that os check?

**Reply by ajaykodam** [reply]
@dlemstra Thank you

**Comment by ajaykodam** [comment]
@dlemstra Do we have any method or properties to wrap the text if the text doesn't fit in the image with the given width/height and it should wrap to the next line?

```
MagickReadSettings readSettings = new()
            {
                FillColor = MagickColors.Black,
                BackgroundColor = MagickColors.White,
                FontPointsize = 12,
                Width = 600,
                Height= 720
            };

using (MagickImage image = new(MagickColors.White, 600, 720))
                {
                    using (MagickImage label = new($"label:{text}", readSettings))
                    {
                        MemoryStream outputMemoryStream = new();
                        image.Composite(label, 10, 10, CompositeOperator.Over);
                        image.Format = MagickFormat.Jpeg;
                        image.Write(outputMemoryStream);
                       return outputMemoryStream.ToArray();
                    }
                }
```

**Reply by dlemstra** [reply]
Can you please create a new discussion topic for follow up questions?

**Reply by ajaykodam** [reply]
Done
